### DISCUSSION TITLE
Is it possible to use dialogs with the Elmish hooks?

### DISCUSSION THREAD

**Discussion by marklam** [fsprojects/Avalonia.FuncUI#250] [Q&A] [answered]
I'm using the Elmish hooks in the component model, but I can't get `dialog.ShowAsync` to work from the update method. 

The non-hook Elmish version has `Cmd.OfAsync`, which is used in the MusicPlayer sample, but the hook version only has `Cmd.ofMsg`. to get the payload for the message, I have to await the `ShowAsync` in the update function. Awaiting the `ShowAsync` seems to hang even if I use the `Dispatcher.UIThread` (this is on Windows BTW). 

Any help would be greatly appreciated!

My update function is `handler` below with `windowParams` partially applied to supply the parent window

```F#
        type WindowParams = { Window : Window }
        type Msg = | PickProjectFile | SetProjectFile of string

        let handler windowParams msg model =
            match msg with
            | PickProjectFile ->
                let saveFileTask =
                    Dispatcher.UIThread.InvokeAsync<string>(
                        System.Func<System.Threading.Tasks.Task<string>>(fun () ->
                            let dialog = SaveFileDialog()
                            dialog.ShowAsync windowParams.Window
                        ))

                model, saveFileTask.Result |> SetProjectFile |> Cmd.ofMsg
```

Thanks.

**Comment by JordanMarr** [comment]
Thanks for posting your issue. The `useElmish` hook is still a very barebones implementation.
However, I am definitely interested in adding proper async support, so I'll take a pass at it.

**Reply by marklam** [reply]
Thanks, that would be great!

**Comment by JordanMarr** [comment]
Also, you may need to add `[<System.STAThread>]` attribute to your main fn.

**Reply by marklam** [reply]
Unfortunately that doesn't seem to fix it.

**Comment by marklam** [comment]
I've added some sample code at https://github.com/marklam/FuncUIFileDialog

It's a standalone solution (using paket - so `dotnet tool restore` before `dotnet build`), but I've also included an fsproj file which could be used to drop the project folder into the `src` folder of the Avalonia.FuncUI repo for easier testing.

**Comment by marklam** [comment]
**Please Ignore this comment, it's fixed (explanation in the replies)**

**As a side note**, the AppBuilder needs a call to `.With(Win32PlatformOptions(UseWgl = true))` otherwise the app throws the exception `eglCreatePbufferFromClientBuffer failed with error EGL_BAD_ATTRIBUTE (0x3004)` at
```text
	Avalonia.OpenGL.dll!Avalonia.OpenGL.Egl.EglPlatformOpenGlInterface.CreatePBufferFromClientBuffer(int bufferType, System.IntPtr handle, int[] attribs) Line 67	C#
 	Avalonia.OpenGL.dll!Avalonia.OpenGL.Angle.AngleWin32EglDisplay.WrapDirect3D11Texture(Avalonia.OpenGL.Egl.EglPlatformOpenGlInterface egl, System.IntPtr handle, int offsetX, int offsetY, int width, int height) Line 92	C#
 	Avalonia.Win32.dll!Avalonia.Win32.WinRT.Composition.WinUiCompositedWindowSurface.CompositionRenderTarget.BeginDraw() Line 75	C#
 	Avalonia.Skia.dll!Avalonia.Skia.GlRenderTarget.BeginRenderingSession() Line 63	C#
 	Avalonia.Skia.dll!Avalonia.Skia.SkiaGpuRenderTarget.CreateDrawingContext(Avalonia.Rendering.IVisualBrushRenderer visualBrushRenderer) Line 27	C#
 	Avalonia.Visuals.dll!Avalonia.Rendering.DeferredRenderer.EnsureDrawingContext(ref Avalonia.Platform.IDrawingContextImpl context) Line 617	C#
 	Avalonia.Visuals.dll!Avalonia.Rendering.DeferredRenderer.UpdateRenderLayersAndConsumeSceneIfNeeded(ref Avalonia.Platform.IDrawingContextImpl context, bool recursiveCall) Line 376	C#
 	Avalonia.Visuals.dll!Avalonia.Rendering.DeferredRenderer.Render(bool forceComposite) Line 323	C#
 	Avalonia.Visuals.dll!Avalonia.Rendering.DeferredRenderer.Avalonia.Rendering.IRenderLoopTask.Render() Line 276	C#
 	Avalonia.Visuals.dll!Avalonia.Rendering.RenderLoop.TimerTick(System.TimeSpan time) Line 145	C#
 	Avalonia.Win32.dll!Avalonia.Win32.WinRT.Composition.WinUICompositorConnection.RunLoopHandler.Invoke(Avalonia.Win32.WinRT.IAsyncAction asyncInfo, Avalonia.Win32.WinRT.AsyncStatus asyncStatus) Line 107	C#
 	Avalonia.Win32.dll!Avalonia.Win32.WinRT.Impl.__MicroComIAsyncActionCompletedHandlerVTable.Invoke(void* this, void* asyncInfo, Avalonia.Win32.WinRT.AsyncStatus asyncStatus) Line 2879	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Avalonia.Win32.dll!Avalonia.Win32.WinRT.Composition.WinUICompositorConnection.RunLoop() Line 131	C#
 	Avalonia.Win32.dll!Avalonia.Win32.WinRT.Composition.WinUICompositorConnection.TryCreateAndRegisterCore.AnonymousMethod__0() Line 81	C#
 	System.Private.CoreLib.dll!System.Threading.Thread.StartCallback() Line 106	C#
```

**The strange thing is** that the call isn't needed if the demo project is built & run within the Avalonia.FuncUI tree, with project references rather than nuget package references.

Is this a known/expected thing, or should I create a separate bug/issue for it?

**Reply by Numpsy** [reply]
I can't comment on the elmish part, but for differences between using the nugets and building in the tree - are you using the same versions of things in both cases? (the sample project you linked looks to be using Avalonia 10, and the latest code in the tree is Avalonia 11, and there are quite a lot of changes over there)

**Reply by marklam** [reply]
Long story short, I found that the OpenGL problem happened because I wasn't asking paket to resolve the dependencies in the right way. 

Short story long, 
- I made a branch from the 0.5.3 release at https://github.com/marklam/Avalonia.FuncUI/tree/needs-use-wgl
- I created a test app using ProjectReference and one using PackageReference and they both worked fine. 
- I then tried using paket instead of PackageRererence, and it didn't work.
- I compared the `*.deps.json` files and found that dependencies-of-dependencies were mismatched versions.
- I removed the paket reference to Avalonia.Desktop and set `strategy:min` on the FuncUI references.
- Then it worked as expected.

I'll update the elmish sample I addd to use PackageReference, to avoid this trip-hazard.

**Comment by JordanMarr** [comment, marked as answer]
The next prerelease will have a completely reworked useElmish hook that will be fully based off of Elmish 4, so it will have all the latest async handlers.

https://github.com/fsprojects/Avalonia.FuncUI/pull/251

**Reply by marklam** [reply]
Can confirm this works now.

Many thanks!
